File: c/matrizes/negativos_matriz.c
/* Problema "negativos_matriz"
 Ler dois números M e N (máximo = 10), e depois ler uma matriz MxN de números inteiros. Em seguida, mostrar na tela somente os números negativos da matriz.
*/

#include <stdio.h>

#define MIN_SIZE 1
#define MAX_SIZE 10

int readSizeOf(char dimension[])
{
  int d;
  printf("Qual a quantidade de %s da matriz (máximo de %d)? ", dimension, MAX_SIZE);

  scanf("%d", &d);

  if (d < MIN_SIZE || d > MAX_SIZE)
  {
    d = MAX_SIZE;
  }

  return d;
}

int main()
{
  int n = readSizeOf("linhas");
  int m = readSizeOf("colunas");

  int matrix[n][m];

  for (int i = 0; i < n; i++)
  {
    for (int j = 0; j < m; j++)
    {
      printf("Elemento [%d, %d]: ", i, j);
      scanf("%d", &matrix[i][j]);
    }
  }

  printf("\n\nVALORES NEGATIVOS: \n");
  for (int i = 0; i < n; i++)
  {
    for (int j = 0; j < m; j++)
    {
      if (matrix[i][j] < 0)
        printf("%d\n", matrix[i][j]);
    }
  }

  return 0;
}

File: python/src/matrizes/negativos_matriz.py
""" Problema "negativos_matriz"
Ler dois números M e N (máximo = 10), e depois ler uma matriz MxN de números inteiros. Em
seguida, mostrar na tela somente os números negativos da matriz.
"""

MIN_SIZE = 1
MAX_SIZE = 10


def read_size_of(dimension: str) -> int:
    size = int(input(f"Qual a quantidade de {dimension} da matriz (mínimo de {MIN_SIZE} e máximo de {MAX_SIZE})? "))

    if not MIN_SIZE < size <= MAX_SIZE:
        size = MIN_SIZE

    return size


def read_matrix(rows: int, columns: int) -> list[list[int]]:
    matrix: list[list[int]] = [[0 for _ in range(columns)] for _ in range(rows)]

    for i in range(len(matrix)):
        print(f"Digite os {len(matrix[0])} elementos da {i + 1}ª linha:")
        for j in range(len(matrix[0])):
            matrix[i][j] = int(input(f"Elemento [{i}, {j}]: "))

    return matrix


def print_negative_elements(matrix: list[list[int]]) -> None:
    print("VALORES NEGATIVOS")
    for i in range(len(matrix)):
        for value in matrix[i]:
            if value < 0:
                print(f"{value}")


def negativos_matriz() -> None:
    rows = read_size_of("linhas")
    columns = read_size_of("colunas")

    matrix: list[list[int]] = read_matrix(rows, columns)

    print_negative_elements(matrix)


if __name__ == "__main__":
    negativos_matriz()
